iopath/common/azure_blob.py

```diff
@@ -0,0 +1,632 @@
+                remote_modified = props["last_modified"]
+                local_modified = datetime.fromtimestamp(
+                    os.path.getmtime(local_path)
+                ).astimezone()
+
+                if remote_modified <= local_modified and not force:
+                    logger.info(
+                        "URL {} was already cached in {}".format(path, local_path)
+                    )
+                    return local_path
+
+            logger.info("Caching {} ...".format(path))
+            tmp = local_path + ".tmp"
+            # clean-up tmp if found, because if tmp exists, it must be a dirty
+            # result of a previously process that didn't cleanup itself.
+            if os.path.isfile(tmp):
+                os.unlink(tmp)
+
+            account, container, blob_path = self._parse_uri(path)
+            client = self._get_client(account)
+            try:
+                blob_stream = client.get_blob_client(
+                    container=container, blob=blob_path
+                ).download_blob()
+
+                with open(tmp, "wb") as f_tmp:
+                    blob_stream.readinto(f_tmp)
+
+                shutil.move(tmp, local_path)
+            finally:
+                try:
+                    os.unlink(tmp)
+                except Exception:
+                    pass
+
+            logger.info("URL {} cached in {}".format(path, local_path))
+            return local_path
+
+    def _open(
+        self,
+        path: str,
+        mode: str = "rb",
+        buffering: int = -1,
+        **kwargs: Any,
+    ) -> Union[IO[str], IO[bytes]]:
+        """
+        Open a stream to a URI, similar to the built-in `open`.
+        Args:
+            path (str): A URI supported by this PathHandler
+            mode (str): Specifies the mode in which the file is opened. It defaults
+                to 'r'.
+            buffering (int): An optional integer used to set the buffering policy.
+                Pass 0 to switch buffering off and an integer >= 1 to indicate the
+                size in bytes of a fixed-size chunk buffer. When no buffering
+                argument is given, the default buffering policy depends on the
+                underlying I/O implementation.
+        Returns:
+            file: a file-like object.
+        """
+        self._check_kwargs(kwargs)
+        assert mode in ("rb", "wb"), "Supported modes: rb, wb"
+
+        chunk_size = buffering if buffering > 0 else DEFAULT_CHUNK_SIZE
+
+        account, container, blob_path = self._parse_uri(path)
+        blob_client = self._get_client(account).get_blob_client(
+            container=container,
+            blob=blob_path,
+        )
+
+        logger.info("Opening blob: path=%s, mode=%s", path, mode)
+
+        if "r" in mode:
+            return AzureBlobReader(blob_client, chunk_size)
+        if "w" in mode:
+            return AzureBlobWriter(blob_client, chunk_size)
```

Comment:
Can you also add a TextIOWrapper to support text mode options, similar to the other handlers, e.g. S3 handler?

Reply:
Sounds good. Please create a separate task to address that.

Reply:
Opened an issue to track: https://github.com/facebookresearch/iopath/issues/18